tephra-core/src/main/java/co/cask/tephra/visibility/VisibilityBarrier.java

```diff
@@ -0,0 +1,121 @@
+ * separate transaction.
+ * If the writer is able to successfully commit the transaction with write barrier, it means that
+ * any in-progress readers that have not seen the change will not be allowed to commit anymore.
+ * In case an in-progress reader commits before the write barrier is committed, then
+ * the writer will get a conflict on commit, and the writer has to retry the write barrier with a new transaction
+ * until it successfully commits.
+ * <p>
+ * Any in-progress readers after the write barrier is committed will get a conflict when they try to commit. This will
+ * force the reader to start a new transaction, and this ensures that the changes made by writer are visible
+ * to the reader.
+ * <p>
+ * Hence a successful commit of a write barrier ensures that any read barrier that successfully commits after
+ * this point onwards would have seen the change.
+ *
+ * <p>
+ * Sample reader code:
+ * <pre>
+ *   <code>
+ * TransactionAware readBarrier = VisibilityBarrier.reader(barrierId);
+ * TransactionContext readTxContext =
+ *        new TransactionContext(new InMemoryTxSystemClient(txManager), readBarrier, table1, table2, ...);
+ * readTxContext.start();
+ *
+ * // do operations using table1, table2, etc.
+ *
+ * // finally commit
+ * try {
+ *   readTxContext.finish();
+ * } catch (TransactionConflictException e) {
+ *   // handle conflict by aborting and starting over with a new transaction using the same barrier object
+ * }
+ *   </code>
+ * </pre>
+ *<p>
+ *
+ * Sample writer code:
+ * <pre>
+ *   <code>
+ * // start transaction
+ * // write change
+ * // commit transaction
+ *
+ * // Now use a write barrier to ensure that all in-progress readers are aware of this change
+ * TransactionAware writeBarrier = VisibilityBarrier.writer(barrierId);
+ * TransactionContext writeTxContext = new TransactionContext(new InMemoryTxSystemClient(txManager), writeBarrier);
+ * writeTxContext.start();
+ * try {
+ *   writeTxContext.finish();
+ * } catch (TransactionConflictException e) {
+ *   // handle conflict by aborting and starting over with a new transaction using the same barrier object
+ * }
+ *   </code>
+ * </pre>
+ *
+ * Note that in this implementation, any reader that starts a transaction after the write is committed,
+ * but before the transaction with write barrier is started will also conflict with the write barrier,
+ * even thought this reader has seen the latest changes.
+ * For now we consider this as an acceptable trade-off to keep the implementation simple.
+ */
+public class VisibilityBarrier {
+  private static final byte[] VISIBILITY_PREFIX = {'v'};
+
+  /**
+   * Returns a read barrier for a given barrier id.
+   *
+   * @param barrierId uniquely identifies the data that this barrier is used to synchronize.
+   *                  If the data is a table cell then this id will be composed of the table name, row key
```

Comment:
``this id will be composed of`` - is this a guideline to the user or will tephra compose the id using this logic?